### ISSUE TITLE
[Feature Request] Ability to discard progress.

### ISSUE THREAD

**Issue by tehguitarist** [rasmuslos/ShelfPlayer#75]
This is somewhat related to this: https://github.com/advplyr/audiobookshelf/issues/2528 as I occasionally have this problem and just use the official app to discard progress a couple of times, but I can see it being helpful elsewhere too.

**Comment by rasmuslos** [comment]
Is the problem that you cannot remove books from continue listing or that you have multiple progress entities for the same item.

Because during the development of ShelfPlayer I had an issue that I would create multiple progress entities for the same item, but we could not find a cause for this, as all endpoints that ShelfPlayer uses for progress tracking call essentially the same functions (https://github.com/advplyr/audiobookshelf/issues/2213)... This should only be relevant if there are completed item in the continue listing row.

If the problem is that you cannot remove items that are in progress the problem is that I did not implement the hide from continue listing thing yet, but it should be quite straightforward.

**Comment by tehguitarist** [comment]
I think the problem is the double progress entities that stay in the continue listening row even though they’re completed. Pressing discard progress twice in the official app, and then marking as finished fixes the issue however.

It’s an ABS issue, so having this would be more of a convenience than anything. I can still fix it in the other app or the web UI easily enough. I guess it’s also an alternative to opening the book and hitting the first chapter to reset progress too.

**Comment by rasmuslos** [comment]
I can look into this, but just out of curiosity, did you start using your current ABS instance before or after the database migration? I think this could be the cause of this issue.

**Comment by tehguitarist** [comment]
My library existed before the migration, but that listening session started after the migration (a couple of weeks back so it would have been 1.8.0 or 1.8.1). Sounds like I might need to open that bug again, I'd rather not re-import my library, though, that should be fairly safe if it reads off the ABS metadata. Like I said, not really your problem, it was only a question if it was easy to implement, I can definitely fix it other ways.

**Comment by rasmuslos** [comment]
I resolved the issue by completing removing abs and that restoring it from a backup.

The thing is I doubt that this bug will be resolved in the near future, and it would be fairly easy for ShelfPlayer to delete most of the bugged sessions so only one remains. I think if only the newest one or the one with the most listening time is kept it should be fine. Is this something you would be interested in?

**Comment by rasmuslos** [comment]
> My library existed before the migration, but that listening session started after the migration (a couple of weeks back so it would have been 1.8.0 or 1.8.1). Sounds like I might need to open that bug again, I'd rather not re-import my library, though, that should be fairly safe if it reads off the ABS metadata. Like I said, not really your problem, it was only a question if it was easy to implement, I can definitely fix it other ways.

Because this is most likely an issue with the database micro, even though the listing sessions are only created afterwards. But this is just a guess, because the bug is really strange....

**Comment by tehguitarist** [comment]
I think your solution, only keeping the newest session would be a good one.
It is an odd bug though. I wonder if a database rebuild would permanently fix it, or only fix the broken sessions.

Thanks for considering this, glad it’s not just me who’s experienced it.

**Comment by rasmuslos** [comment]
> I think your solution, only keeping the newest session would be a good one. It is an odd bug though. I wonder if a database rebuild would permanently fix it, or only fix the broken sessions.
> 
> Thanks for considering this, glad it’s not just me who’s experienced it.

Purging and restoring your ABS instance will resolve the issue and prevent further issues. But it is a really odd bug. Can you give me your E-Mail address so that I can add a build to your TestFlight to see if the implementation works? I cannot really test it myself.

Glad I can help!

**Comment by tehguitarist** [comment]
No worries, I don't currently have the issue, but can let you know next time I do? I'm towards the end of a book so it may hit again very soon. Happy to send my email regardless if you'd like though. I use a different email for the AppStore than I do for iCloud so I'll do a search to see which one is likely linked to my TestFlight (likely the appstore email).

**Comment by rasmuslos** [comment]
Yes, this should be the one. Can you send me an E-Mail to git@rfk.io?

**Comment by tehguitarist** [comment]
Just shot you an email now. The App Store email is different to the one I sent it from though.

**Comment by tehguitarist** [comment]
Just tested the build, I opted into removing duplicate sessions in the settings, but nothing seems to have changed. The most recent book did fault, and clearing listen history and hitting mark as read didn't change anything.

**Comment by rasmuslos** [comment]
The problem is, that i don't have a way of debugging this. I can look at my implementation again, maybe I will find a flaw, but this issue is sadly not reproducible... The only thing that comes to mind right now is, did you enable the feature in the settings app in the "advanced" section?

**Comment by tehguitarist** [comment]
Yeah I did enable it. It’s all good, I’ll probably refresh my library to fix it once and for all when I get home from being away. Not a show stopper. Thanks so much again for your time.

**Comment by sn0opy** [comment]
@tehguitarist did you manage to fix it eventually? Running into the same issue here.

**Comment by tehguitarist** [comment]
I ended up trashing my library and reimporting (with abs metadata as first priority). It actually found a few file problems I had and didn’t realise so it wasn’t all for nothing.

**Comment by rasmuslos** [comment]
> did you manage to fix it eventually? Running into the same issue here.

If you want to avoid the issue in the future make sure to download the items you are listening to. I am still working out why this happens, because the code that updates the progress is the same on the server so its really weird that this happens.

**Comment by sn0opy** [comment]
Maybe one more thing, that might narrow it down: it seems to only happen if the app is in the background while finishing an episode. Not sure if the way it plays audio in the background is different or other things are not happen properly while being in the background.
